src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp

```diff
@@ -323,16 +364,23 @@ template <typename T>
 bool ShenandoahBarrierSet::AccessBarrier<decorators, BarrierSetT>::oop_arraycopy_in_heap(arrayOop src_obj, size_t src_offset_in_bytes, T* src_raw,
                                                                                          arrayOop dst_obj, size_t dst_offset_in_bytes, T* dst_raw,
                                                                                          size_t length) {
+  T* src = arrayOopDesc::obj_offset_to_raw(src_obj, src_offset_in_bytes, src_raw);
+  T* dst = arrayOopDesc::obj_offset_to_raw(dst_obj, dst_offset_in_bytes, dst_raw);
+
   ShenandoahBarrierSet* bs = ShenandoahBarrierSet::barrier_set();
-  bs->arraycopy_barrier(arrayOopDesc::obj_offset_to_raw(src_obj, src_offset_in_bytes, src_raw),
-                        arrayOopDesc::obj_offset_to_raw(dst_obj, dst_offset_in_bytes, dst_raw),
-                        length);
-  return Raw::oop_arraycopy_in_heap(src_obj, src_offset_in_bytes, src_raw, dst_obj, dst_offset_in_bytes, dst_raw, length);
+  bs->arraycopy_barrier(src, dst, length);
+  bool result = Raw::oop_arraycopy_in_heap(src_obj, src_offset_in_bytes, src_raw, dst_obj, dst_offset_in_bytes, dst_raw, length);
+  if (ShenandoahCardBarrier) {
+    bs->write_ref_array((HeapWord*) dst, length);
```

Comment:
Um. So we are here, where `T` might mean either `oop` or `narrowOop`. Yet, we are calling `write_ref_array` with `HeapWord*` and `length`, which is here a _count_ of either `oop` or `narrowOop`-s. So for `narrowOop`, which is smaller than `HeapWord`, we call `write_ref_array` with _double_ the length? 

Note how `ShenandoahBarrierSet::arraycopy_barrier` carries `T* src, size_t count` all the way down to `ShenandoahBarrierSet::arraycopy_work` to avoid this problem. I think `write_ref_array` should really carry overloads for `write_ref_array(oop* src, size_t count)` and `write_ref_array(narrowOop* src, size_t count)`.

Reply:
@kdnilsen will take a closer look.

Reply:
Tracking this in https://bugs.openjdk.org/browse/JDK-8339615

Reply:
Inside of write_ref_array(), we calculate the end of the array using:
```
  HeapWord* end = (HeapWord*)((char*) start + (count * heapOopSize));
```
Then we mark between array start and array end as dirty.  Though this code as written is a "bit" clumsy, I believe it is correct.

Reply:
Am willing to try cleaning up the code, but am concerned this introduces risk of introducing regressions.  I see that the same pattern appears in the generic ModRefBarrierSet::write_ref_array() and in ModRefBarrierSet::AccessBarrier::oop_arraycopy_in_heap().

Reply:
All right.